### DISCUSSION TITLE
Compress/decompress from and to memory rather than file

### DISCUSSION THREAD

**Discussion by cwiggins999** [gildas-lormeau/zip.js#435] [Q&A] [answered]
Can zip.js be used to compress/decompress from and to a variable (string I guess) rather than a file? I'd like to be able to send a compressed buffer to a server, as well as decompress that buffer when received.

**Comment by gildas-lormeau** [comment, marked as answer]
You should not use a string to store binary data. Depending on the size of the data, you can use an instance of `ReadableStream` (see https://github.com/gildas-lormeau/zip.js/blob/master/tests/all/test-readable-stream.js), or `BlobReader` (see https://github.com/gildas-lormeau/zip.js/blob/a64c5238c6bd20b64b6aa32551e80347289ac5fb/tests/all/test-blob.js), or `Uint8ArrayReader` (see https://github.com/gildas-lormeau/zip.js/blob/master/tests/all/test-array.js).

**Reply by cwiggins999** [reply]
Thanks - very helpful! Do these methods support encryption/password protection? Our app running on Electron and Capacitor, and we already have the FileSystem read/writes, but would like to encrypt/decrypt as well as compress.

**Reply by gildas-lormeau** [reply]
Yes, as stated in the documentation, you can pass it in the `password` property of the second parameter of `ZipWriter` and `Entry#getData()`, see for example:

https://github.com/gildas-lormeau/zip.js/blob/eb89a9c2448bd13637a86cb60c8784c33b5ae43f/tests/all/test-crypto.js#L14

https://github.com/gildas-lormeau/zip.js/blob/eb89a9c2448bd13637a86cb60c8784c33b5ae43f/tests/all/test-crypto.js#L33

**Reply by cwiggins999** [reply]
I saw that for the ZipWriter, but I would like to just compress/encrypt as a "Blob" rather than creating a zip file, and use my existing file i/o routines. Is that possible?

**Reply by gildas-lormeau** [reply]
You mean you don't want to use zip.js?

**Comment by cwiggins999** [comment]
I want to compress/decompress data and also gain encryption, but I want to read/write files using my existing routines - since in Capacitor (mobile app) I can't use anything but their native read/write routines. So I want to encrypt/compress JSON data into a buffer and write that buffer out/read it back in myself. So I hoped I could use some functionality of zip.js and gain encryption. (So far I've used pako, which works well but doesn't support encryption.)

Thanks!

**Reply by gildas-lormeau** [reply]
The encryption feature I showed as example does not prevent to read/write files with your existing routines. Otherwise, you should use the native API for this, see https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/encrypt and https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/decrypt for example.

**Reply by cwiggins999** [reply]
Do you mean "does not allow"? 

Thanks for answering my questions - even if this isn't the library for me!

**Reply by gildas-lormeau** [reply]
No, I mean "does not prevent". You can write your own Reader/Writer or use Web Streams. Writing/reading encrypted zip files in zip.js does not prevent to do that. The encryption is handled separately and transparently for you.

**Comment by cwiggins999** [comment]
Got it - and I put together some test code in both my environments - Electron and Capacitor - and looks like this is going to work well. Thanks again for the help!!

**Reply by gildas-lormeau** [reply]
You're welcome !
